GTL: pass options through preventing double call to reporting backend

martinpovolny opened the pull request:
Taking over https://github.com/ManageIQ/manageiq-ui-classic/pull/2335

martinpovolny commented:
@AparnaKarve : in `spec/controllers/ops_controller/ops_rbac_spec.rb` in this PR there's o hint of how we can write tests for the display of data in GTLs:

There are 2 parts:
  1. tests the flow up to the place where `get_view` is called (with `false`)
  2. tests the flow from the `get_view` called from `/report_data` (with `true`)

Far from perfect, but I wanted to keep the tests at least.

AparnaKarve commented:
@martinpovolny Looks like the PR is currently addressing this - `Additional options for report_data`, but the other issue - `Do not fetch multiple times view from DB` still seems open. Would it be addressed in the next phase of this PR?

As far as the tests go, would it be possible for `get_view` to populate an instance variable that contains the count of records (or rows) that will be *actually* rendered.
This instance variable may not have a role to play in the GTL rendering, but it would merely exist to assist us in our spec where we want to make sure that only `x` number of rows are being rendered in a nested display.

martinpovolny commented:
> but the other issue - Do not fetch multiple times view from DB still seems open. Would it be addressed in the next phase of this PR?

it is actually addressed in this PR.

We discussed a lot with @karelhala today and there are some problems that need to be addressed. Mostly with the fact that many calls to `get_node_infor` actually set `@view` and `@pages` as a side effect :-(

Saying that I think it' also an answer to your question:

> As far as the tests go, would it be possible for get_view to populate an instance variable that contains the count of records (or rows) that will be actually rendered.

I conside that a bad practice. I will try to make sure that we avoid creating any new instance variables in controller unless absolutly necessary.

The presence of such variables and the side effects this (enables and) creates is the reason why any cleanup takes soooooo long.

karelhala commented:
> Looks like the PR is currently addressing this - Additional options for report_data, but the other issue - Do not fetch multiple times view from DB still seems open. Would it be addressed in the next phase of this PR?

It should not fetch data multiple times from DB right now, since we added condition which is passed only when rendering data through report_data [application_controller.rb#L1618](https://github.com/martinpovolny/manageiq-ui-classic/blob/70495e8bfcb031e72b86b5fa5a2dc4d7c2d042a7/app/controllers/application_controller.rb#L1618), of course when changing page DB is called once again. But issue which was raised that when rendering screen with data and GTL should fetch data only when GTL asks for them should be already fixed.

AparnaKarve commented:
> I conside that a bad practice. I will try to make sure that we avoid creating any new instance variables in controller unless absolutly necessary.

True, I'm not a fan of ivars in general - but the only reason I suggested that option was to have a quick spec in there and not having to go through those frustrating regressions again.

martinpovolny commented:
@AparnaKarve : Please, see my comment:

https://github.com/ManageIQ/manageiq-ui-classic/pull/2383#issuecomment-336800557

yes, we need to figure out how to make this testeable.

But regression there will be. Because the test coverage is low now. We need at least a simple spec for each of the lists and nested lists, but that is a load of work.

AparnaKarve commented:
@martinpovolny I think we are just going around in circles.

But yes, the main thing that we both agree on is that  - we *do* need a simple spec for  each of the lists and nested lists, and yes, that sure is a load of work.

I have not looked into the GTL component that closely in `ui-components` yet, but do you think  that it would it be possible to write this spec from `ui-components`?

martinpovolny commented:
> I have not looked into the GTL component that closely in ui-components yet, but do you think that it would it be possible to write this spec from ui-components?

No.

In ui-components the best we can get is unit tests for the component. What we need on ui-classic side for each list is something like https://github.com/ManageIQ/manageiq-ui-classic/pull/2383#issuecomment-336800557

Something like that combined with `assert_nested_list` that we have in our specs. But for every single list.

martinpovolny commented:
One of BZ that will be resolved by this:
https://bugzilla.redhat.com/show_bug.cgi?id=1501376

We have 2 identical requests: `http://localhost:3000/ops/report_data`  with payload `{"active_tree":"rbac_tree","explorer":true}`

Both call reach the line `MiqReport.load_from_view_options(db, current_user, options, db_view_yaml_cache)`

Once `db` is set to `"MiqGroup"` and once it's empty. So we need to pass the model through and stop using crazy hacks to get it set.

I'd say that this and similar BZ need to wait until we merge this. We need to collect the problems and make sure we fix all, but should not try to "fix the hacks".

martinpovolny commented:
We have one more issue to fix: The where_clauses we removed, thx @tumido, but there are also `:conditions` :-(

```
devil - [~/Projects/manageiq-ui-classic] (gtl_pass_through)$ grep -r :conditions app/controllers/
app/controllers/ops_controller/diagnostics.rb:      @view, @pages = get_view(MiqServer, :conditions => ["zone_id=?", zone.id]) # Get the records (into a view) and the paginator
app/controllers/ops_controller/diagnostics.rb:    @view, @pages = get_view(MiqWorker, :conditions => ["(miq_server_id = ? and status IN (?))", @sb[:selected_server_id], status], :all_pages => true) # Get the records (into a view) and the paginator
app/controllers/ops_controller/settings/schedules.rb:    @view, @pages = get_view(MiqSchedule, :conditions => ["prod_default!=? or prod_default IS NULL And adhoc IS NULL", "system"]) # Get the records (into a view) and the paginator
app/controllers/pxe_controller/pxe_customization_templates.rb:      @view, @pages = get_view(CustomizationTemplate, :conditions => ["pxe_image_type_id=?", from_cid(pxe_img_id)]) # Get the records (into a view) and the paginator
app/controllers/pxe_controller/pxe_customization_templates.rb:      @view, @pages = get_view(CustomizationTemplate, :conditions => ["system=?", true])  # Get the records (into a view) and the paginator
app/controllers/mixins/explorer_show.rb:        show_details(klass, :association => association, :conditions => conditions)
app/controllers/mixins/explorer_show.rb:      conditions  = options[:conditions]
app/controllers/mixins/explorer_show.rb:                               :conditions  => conditions,
app/controllers/miq_policy_controller/policies.rb:      mems = @edit[:new][:conditions].invert # Get the ids from the member list box
app/controllers/miq_policy_controller/policies.rb:    handle_selection_buttons(:conditions)
app/controllers/miq_policy_controller/policies.rb:    peca_get_all('policy', -> { get_view(MiqPolicy, :conditions => ["mode = ? and towhat = ?", @sb[:mode].downcase, @sb[:nodeid].camelize]) })
app/controllers/miq_policy_controller/policies.rb:      @edit[:new][:conditions] = {}
app/controllers/miq_policy_controller/policies.rb:      conditions.each { |c| @edit[:new][:conditions][c.description] = c.id }   # Build a hash for the members list box
app/controllers/miq_policy_controller/policies.rb:      @edit[:new][:conditions].each_key { |key| @edit[:choices].delete(key) }  # Remove any choices that are in the members list box
app/controllers/chargeback_controller.rb:    @view, @pages = get_view(ChargebackRate, :conditions => ["rate_type=?", x_node.split('-').last])  # Get the records (into a view) and the paginator
app/controllers/infra_networking_controller.rb:    conditions  = options[:conditions]
app/controllers/infra_networking_controller.rb:                             :conditions  => conditions,
app/controllers/application_controller.rb:    view.conditions = options[:conditions] # Get passed in conditions (i.e. tasks date filters)
app/controllers/miq_task_controller.rb:    @view, @pages = get_view(MiqTask, :conditions => tasks_condition(@tasks_options[@tabform]))
app/controllers/miq_task_controller.rb:                {:date => task.updated_on, :conditions => @tasks_options[@tabform].inspect}
app/controllers/miq_request_controller.rb:      @view, @pages = get_view(MiqProvision, :conditions => ["miq_request_id=?", @miq_request.id])  # Get all requests
app/controllers/report_controller/schedules.rb:      @view, @pages = get_view(MiqSchedule, :conditions => ["towhat=?", "MiqReport"]) # Get the records (into a view) and the paginator
app/controllers/report_controller/schedules.rb:      @view, @pages = get_view(MiqSchedule, :conditions => ["towhat=? AND userid=?", "MiqReport", session[:userid]])  # Get the records (into a view) and the paginator
app/controllers/vm_common.rb:      @record_pages, @records = paginate(:vms, :per_page => @items_per_page, :order => @col_names[get_sort_col] + " " + @sortdir, :conditions => ["id IN (?)", selected])
app/controllers/miq_ae_customization_controller/old_dialogs.rb:      @view, @pages = get_view(MiqDialog, :conditions => ["dialog_type=?", x_node.split('_').last]) # Get the records (into a view) and the paginator
```

So the next step is to convert these to scopes...

martinpovolny commented:
help welcome reviewing and testing not only this PR but also:

https://github.com/ManageIQ/manageiq-ui-classic/pull/2513
https://github.com/ManageIQ/manageiq/pull/16303

AparnaKarve commented:
@martinpovolny I will verify in a couple of minutes, but in the meantime, can you update the spec like this -
```ruby
it 'displays associated storage_managers' do
 FactoryGirl.create(:ems_storage, :name => 'abc', :type =>  "ManageIQ::Providers::Amazon::StorageManager::Ebs", :parent_ems_id => @ems.id)
 FactoryGirl.create(:ems_storage, :name => 'xyz', :type =>  "ManageIQ::Providers::Amazon::StorageManager::Ebs", :parent_ems_id => @ems.id)
 get :show, :params => { :display => "storage_managers", :id => @ems.id, :format => :js }
 is_expected.to render_template(:partial => "layouts/angular/_gtl")
 expect(response.status).to eq(200)
end
```

The `render_template` part will fail if the yaml loading did not go through properly.
Thanks.

dclarizio commented:
@h-kataria can you verify this fix?  Thx, Dan

AparnaKarve commented:
@martinpovolny Getting a syntax error -

```
F, [2017-10-26T18:56:44.033087 #49523] FATAL -- : SyntaxError (~/manageiq/plugins/manageiq-ui-classic/app/controllers/mixins/generic_show_mixin.rb:184: syntax error, unexpected ':', expecting ')'
e(options.slice(:association :parent_method :where_clause :n
                              ^
~/manageiq/plugins/manageiq-ui-classic/app/controllers/mixins/generic_show_mixin.rb:184: syntax error, unexpected ':', expecting keyword_end
:parent_method :where_clause :named_scope))
                              ^):
F, [2017-10-26T18:56:44.033169 #49523] FATAL -- :   
F, [2017-10-26T18:56:44.033224 #49523] FATAL -- : plugins/manageiq-ui-classic/app/controllers/mixins/generic_show_mixin.rb:184: syntax error, unexpected ':', expecting ')'
e(options.slice(:association :parent_method :where_clause :n
                              ^
plugins/manageiq-ui-classic/app/controllers/mixins/generic_show_mixin.rb:184: syntax error, unexpected ':', expecting keyword_end
:parent_method :where_clause :named_scope))
```

martinpovolny commented:
Yeah, I have broken in in the last "style" commit. Did you get around it? Now it's fixed.

martinpovolny commented:
@AparnaKarve : do you have a commit that I can cherry-pick for the spec?

AparnaKarve commented:
>  do you have a commit that I can cherry-pick for the spec?

Yes. Apply this - https://github.com/ManageIQ/manageiq-ui-classic/pull/2378/commits/8f0b6600d7d650f6b01bb9a046868bda8db05d98 on top of the existing one.

AparnaKarve commented:
@martinpovolny I confirmed that the fix works

martinpovolny commented:
So the travis finally passed. I really need a review and testing here. Ping @mzazrivec, @karelhala, @AparnaKarve, @h-kataria.

martinpovolny commented:
Now rebase, travis running once again.

karelhala commented on app/controllers/application_controller.rb:
```diff
@@ -435,16 +435,16 @@ def init_report_data(controller_name)
+    # if params[:active_tree] && defined? get_node_info
+    #   node_info = get_node_info(x_node, false)
+    #   options.merge!(node_info) if node_info.kind_of?(Hash)
+    # end
```
No need to use commented out code.

h-kataria commented:
merging, any other changes/fixes can be done in follow up PR
